File: fb/109.Convert-Sorted-List-to-Binary-Search-Tree.cpp
// Time O(n)
// Space O(logn)
class Solution {
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if (!head) return nullptr;
        ListNode* node = head;
        int count = 0;
        while (node) {
            node = node->next;
            count++;
        }
        cur = head;
        return sortedListToBST(0, count - 1);
    }
    
private:
    ListNode* cur;
    
    TreeNode* sortedListToBST(int lo, int hi) {
        if (lo > hi) return nullptr;
        int mid = lo + (hi - lo)/ 2;
        TreeNode* left = sortedListToBST(lo, mid - 1);
        TreeNode* root = new TreeNode(cur->val);
        cur = cur->next;
        TreeNode* right = sortedListToBST(mid + 1, hi);
        root->left = left;
        root->right = right;
        return root;
    }
};

File: fb/109.Convert-Sorted-List-to-Binary-Search-Tree.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countSize(self, head):
        count = 0
        while head:
            head = head.next
            count += 1
        return count
    
    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        def recursion(left, right):
            nonlocal head
            if left > right:
                return None
            
            mid = left + (right-left)//2
            left = recursion(left, mid-1)
            
            root = TreeNode(head.val)
            root.left = left
            
            head = head.next
            
            root.right = recursion(mid+1, right)
            return root
        
        return recursion(0, self.countSize(head)-1)
